src/Axe.Windows.Telemetry/Logger.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Axe.Windows.Telemetry
+{
+    /// <summary>
+    /// Wraps telemetry logging and sends it to 
+    /// a supplied IAxeWindowsTelemetry implementation
+    /// </summary>
+    public static class Logger
+    {
+        private static IAxeWindowsTelemetry _telemetry = null;
+        private static readonly Object _lockObject = new Object();
+
+        /// <summary>
+        /// Whether or not telemetry is enabled. Exposed to allow callers who do lots of
+        /// work to short-circuit their processing when telemetry is disabled
+        /// </summary>
+        public static bool IsEnabled => _telemetry != null && _telemetry.IsEnabled;
+
+        /// <summary>
+        /// Sets the telemetry sink for Axe.Windows.
+        /// <param name="telemetry">Telemetry from Axe.Windows will be forwarded to this object</param>
+        public static void SetTelemetrySink(IAxeWindowsTelemetry telemetry)
+        {
+            lock (_lockObject)
+            {
+                _telemetry = telemetry;
+            }
+        }
+
+        /// <summary>
+        /// Publishes event with single property/value pair to the current telemetry pipeline
+        /// </summary>
+        /// <param name="action"></param>
+        /// <param name="property"></param>
+        /// <param name="value"></param>
+        public static void PublishTelemetryEvent(TelemetryAction action, TelemetryProperty property, string value)
+        {
+            // Check IsEnabled because ToString on enums is expensive
+            if (!IsEnabled) return;
+
+            PublishTelemetryEvent(action, new Dictionary<TelemetryProperty, string>
+                    {
+                        { property, value }
+                    });
+        }
+
+        /// <summary>
+        /// Publishes event to the current telemetry pipeline
+        /// </summary>
+        /// <param name="action">The action being recorded</param>
+        /// <param name="propertyBag">Associated property bag--this may be null</param>
+        public static void PublishTelemetryEvent(TelemetryAction action, IReadOnlyDictionary<TelemetryProperty, string> propertyBag = null)
+        {
+            // Check IsEnabled because ToString on enums is expensive
+            if (!IsEnabled) return;
+
+            lock (_lockObject)
```

Comment:
>lock [](start = 12, length = 4)

This lock seems unnecessary. _telemetry will be a reference, which is your register size, so it's an atomic operation to change it. If it's null, then the existing code will cover it. If it's not null, then it's guaranteed to be valid. The only case I can think of is Axe.Windows-generated telemetry that occurs near simultaneously with enabling or disabling the telemetry, and then those case seem equivalent to calls that arrive just before or just after the lock exits--some will get through, others won't.

Reply:
When you say the existing code, do you mean the existing null checks?